### PR TITLE
C++: Add remote flow sources via models

### PR DESCRIPTION
**Pull request by rdmarsh2**
This adds a new `RemoteFlowSource` class for IR flow, based on a new model interface. I've included flow source models for `fread` and `gets`, since they had model classes already.

This is consistent with the direction the C++ library is moving for flow sinks such as allocation sizes and format arguments, but rather different from the Java and C# versions of `RemoteFlowSource`. I've also opened #3274 showing a version that's closer to the Java and C# versions.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/security/FlowSources.qll [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import semmle.code.cpp.ir.dataflow.DataFlow
+private import semmle.code.cpp.ir.IR
+import semmle.code.cpp.models.interfaces.FlowSource
+
+/** A data flow source of remote user input. */
+abstract class RemoteFlowSource extends DataFlow::Node {
+}
+
+class FileDescriptorTaintedReturnSource extends RemoteFlowSource {
+  FileDescriptorTaintedReturnSource() {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/security/FlowSources.qll:15]:
```suggestion
private class TaintedReturnSource extends RemoteFlowSource {
  TaintedReturnSource() {
```

This is not restricted to just file descriptors, right?

File: cpp/ql/src/semmle/code/cpp/security/FlowSources.qll [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      instr.getStaticCallTarget() = func and
+      func.hasFlowSource(output) and
+      output.isReturnValue()
+    )
+  }
+}
+
+class FileTaintedParameterSource extends RemoteFlowSource {
+  FileTaintedParameterSource() {
+    exists(RemoteFlowFunction func, ReadSideEffectInstruction instr, FunctionOutput output |
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/security/FlowSources.qll:27]:
```suggestion
    exists(RemoteFlowFunction func, WriteSideEffectInstruction instr, FunctionOutput output |
```

File: cpp/ql/src/semmle/code/cpp/security/FlowSources.qll [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      asInstruction() = instr and
+      instr.getStaticCallTarget() = func and
+      func.hasFlowSource(output) and
+      output.isReturnValue()
+    )
+  }
+}
+
+class FileTaintedParameterSource extends RemoteFlowSource {
+  FileTaintedParameterSource() {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/security/FlowSources.qll:26]:
```suggestion
private class TaintedParameterSource extends RemoteFlowSource {
  TaintedParameterSource() {
```

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+import cpp
+import FunctionInputsAndOutputs
+import semmle.code.cpp.models.Models
+
+/**
+ * A library function which returns data read from a network connection.
+ */
+abstract class RemoteFlowFunction extends Function {
+  abstract predicate hasFlowSource(FunctionOutput output);
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll:17]: If we later add a `LocalFlowFunction`, like C# has `LocalFlowSource`, the name `hasFlowSource` becomes ambiguous. I suggest renaming to `hasRemoteFlowSource`.

Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll:17]: Would it be any problem to add a `string` parameter to hold the textual description of the flow type, like you did in #3274 to be compatible with the other languages?

File: cpp/ql/src/semmle/code/cpp/security/FlowSources.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/security/FlowSources.qll:15]: It's not.

File: cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll (hunk 4 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll]: Done

Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/FlowSource.qll]: Done

File: cpp/ql/src/semmle/code/cpp/models/implementations/Fread.qll [hunk 5]
```diff
@@ -13,7 +13,8 @@ class Fread extends AliasFunction, RemoteFlowFunction {
 
   override predicate parameterIsAlwaysReturned(int n) { none() }
 
-  override predicate hasFlowSource(FunctionOutput output) {
-    output.isParameterDeref(0)
+  override predicate hasRemoteFlowSource(FunctionOutput output, string description) {
+    output.isParameterDeref(0) and
+    description = "String read by " + this.getName()
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Fread.qll:18]: This description looks good enough to me. If we're not planning to provide any more detail, then it doesn't seem necessary to have a `description` parameter on `hasRemoteFlowSource`. The same information is available in the `Tainted{Parameter,Return}Source` classes, and it should be less work to construct the description there than to repeat it for every `RemoteFlowFunction`.

Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Fread.qll:18]: I was expecting that we would provide more detail when it's useful, but `fread` and `gets` are the least-specific sources we'll deal with.

**Pull request merged**